### DISCUSSION TITLE
cmp autocompletion, how to trigger?

### DISCUSSION THREAD

**Discussion by tbaumann** [nix-community/nixvim#928] [Q&A] [answered]
`nvim-cmp` seems to have different out of the box experience. The upstream docu seems to suggest that it will just work when enabled.
But with nixvim apparently nothing works out of the box at least. I get no autocomplete box.

I tried these settings, one by one but that doesn't seem to help either
```nix
      nvim-cmp.mapping = {
        "<C-Space>" = "cmp.mapping.complete()";
        "<C-d>" = "cmp.mapping.scroll_docs(-4)";
        "<C-e>" = "cmp.mapping.close()";
        "<C-f>" = "cmp.mapping.scroll_docs(4)";
        "<CR>" = "cmp.mapping.confirm({ select = true })";
        "<S-Tab>" = {
          action = "cmp.mapping.select_prev_item()";
          modes = [
            "i"
            "s"
          ];
        };
        "<Tab>" = {
          action = "cmp.mapping.select_next_item()";
          modes = [
            "i"
            "s"
          ];
        };
      };
      nvim-cmp.completion.autocomplete = [ "TextChanged" ];
```
Hitting Control+Space should have at least make it pop up, but again nothing.

I'm feeling a bit dumb. I have loads of sources configured. And those which make sense for the file-type show up when I type `:CmpStatus`

My cmp sources
```nix
      cmp-calc.enable = true;
      cmp-clippy.enable = true;
      cmp-cmdline.enable = true;
      cmp-fuzzy-buffer.enable = true;
      cmp-fuzzy-path.enable = true;
      cmp-vim-lsp.enable = true;
      cmp-nvim-lsp.enable = true;
      cmp-path.enable = true;
      copilot-cmp.enable = true;
      copilot-lua.enable = true;
      copilot-lua.panel.enabled = false;
      copilot-lua.suggestion.enabled = false;
```

**Comment by GaetanLepage** [comment]
This is not how you should set up your sources. Instead, you should list them in `plugins.nvim-cmp.sources`.
Here is an example: https://github.com/GaetanLepage/dotfiles/blob/ac9da05c94268437f6cdc23c4cf0a48a74c11c70/home/modules/tui/neovim/completion.nix#L46-L57

**Reply by GaetanLepage** [reply]
This is enough: you don't need to manually enable the sources (`copilot-lua.enable = true;`), this will be done automatically.

**Reply by tbaumann** [reply]
Thanks, I will play around with that template. I see some nice patterns there.

But wow, a lot more complex than I would  have assumed. I would have really expected that cmp sources somehow auto-register :D

**Reply by GaetanLepage** [reply, marked as answer]
It's kind of the other way around: you tell what sources you want and we (nixvim) automatically enable the relevant plugin.

**Reply by refaelsh** [reply]
> This is not how you should set up your sources. Instead, you should list them in `plugins.nvim-cmp.sources`. Here is an example: https://github.com/GaetanLepage/dotfiles/blob/ac9da05c94268437f6cdc23c4cf0a48a74c11c70/home/modules/tui/neovim/completion.nix#L46-L57

1. The `mapping` option you use there - its not longer exists in the docs.
2. The same about `sources`.

What am I missing here?

**Reply by tbaumann** [reply]
It's life with nixvim. I think I will stop using it for now and wait until the dust settled.

It breaks every two weeks. I can't live like that.

**Reply by khaneliman** [reply]
> > This is not how you should set up your sources. Instead, you should list them in `plugins.nvim-cmp.sources`. Here is an example: https://github.com/GaetanLepage/dotfiles/blob/ac9da05c94268437f6cdc23c4cf0a48a74c11c70/home/modules/tui/neovim/completion.nix#L46-L57
> 
>     1. The `mapping` option you use there - its not longer exists in the docs.
> 
>     2. The same about `sources`.
> 
> 
> What am I missing here?

It was renamed to be under `cmp.settings.mapping` and `cmp.settings.sources`. Has been working fine for me for a long time https://github.com/khaneliman/khanelivim/blob/f77a4bae597efc7c907021bb83b545bba1a61af7/packages/khanelivim/plugins/cmp.nix#L26

**Reply by MattSturgeon** [reply]
> It's life with nixvim. I think I will stop using it for now and wait until the dust settled.

If you ever run into breaking changes that don't come with adequate instructions for migrating _or_ have a built-in migration alias _please_ report that as a bug.

While we can't do it in every case, we strive to make any refactoring change as painless as possible. This is just the reality of `main` being both our dev-branch and the branch that tracks `nixos-unstable`.

Incidentally, nixos-unstable _also_ often refactor things to add support for new features or allow more "freeform" or "structured config" definitions.

Unfortunately our docs aren't smart enough to pick up on aliases/renames, so links to the old option names will just 404, but building an outdated config should still work and should print warnings stating exactly what has changed.

**Reply by refaelsh** [reply]
> This is enough: you don't need to manually enable the sources (`copilot-lua.enable = true;`), this will be done automatically.

I think, that this should be added to the docs. Its not clear at all that only `sources` is needed and there is no need for `copilot-lua.enable = true;` and the likes.

**Reply by khaneliman** [reply]
> > This is enough: you don't need to manually enable the sources (`copilot-lua.enable = true;`), this will be done automatically.
> 
> I think, that this should be added to the docs. Its not clear at all that only `sources` is needed and there is no need for `copilot-lua.enable = true;` and the likes.

https://nix-community.github.io/nixvim/plugins/cmp/settings/index.html#pluginscmpsettingssources 

> WARNING: If plugins.cmp.autoEnableSources Nixivm will automatically enable the corresponding source plugins. This will work only when this option is set to a list. If you use a raw lua string, you will need to explicitly enable the relevant source plugins in your nixvim configuration.

**Reply by MattSturgeon** [reply]
@refaelsh I've opened #2094

**Reply by refaelsh** [reply]
> If plugins.cmp.autoEnableSources Nixivm will automatically enable the corresponding source plugins.

Oh! Now I get it. I saw that line, but I did not understand it. Now I do. Thank you :-)

**Reply by refaelsh** [reply]
> @refaelsh I've opened #2094

Great. Thanks.

**Comment by refaelsh** [comment]
> It was renamed to be under cmp.settings.mapping and cmp.settings.sources.

I still do not see it in the docs. Here is a screenshot:
![2024-08-26_22-47](https://github.com/user-attachments/assets/dd4232c6-b593-4d3f-8e63-75e1e77ec642)

**Reply by khaneliman** [reply]
That's only the nested options that have children. 
![image](https://github.com/user-attachments/assets/cb4bda82-7662-436a-addd-6640d7c68fe3)

**Reply by khaneliman** [reply]
https://nix-community.github.io/nixvim/plugins/cmp/settings/index.html#pluginscmpsettingsmapping and https://nix-community.github.io/nixvim/plugins/cmp/settings/index.html#pluginscmpsettingssources

**Reply by refaelsh** [reply]
> That's only the nested options that have children. ![image](https://private-user-images.githubusercontent.com/1778670/361556758-cb4bda82-7662-436a-addd-6640d7c68fe3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ3MDIyMTMsIm5iZiI6MTcyNDcwMTkxMywicGF0aCI6Ii8xNzc4NjcwLzM2MTU1Njc1OC1jYjRiZGE4Mi03NjYyLTQzNmEtYWRkZC02NjQwZDdjNjhmZTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjZUMTk1MTUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTk3YzEwYjI3N2M5N2M0NmE5ZThkZmQ2YzQxNzFlZmQyNzljODRhMmFhYjhjOGFiOGY0YWEyNDI0ZDk5MTA1NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.wlQ9ne5kg5v9H4uadu4IpvBa6NjVGIr6MBOIFPKguHA)

And now I feel stupid :-)

Thank you for your answer.
